File: stripe_webhook.py
import stripe
import uvicorn
from fastapi import FastAPI, Request, Header
import config
from notion_handler import update_payment_status  # 後で実装

app = FastAPI()

# StripeのWebhook Secret
stripe.api_key = config.STRIPE_SECRET_KEY
endpoint_secret = config.STRIPE_WEBHOOK_SECRET

@app.post("/webhook")
async def stripe_webhook(
    request: Request,
    stripe_signature: str = Header(None)
):
    payload = await request.body()
    sig_header = stripe_signature

    try:
        event = stripe.Webhook.construct_event(
            payload, sig_header, endpoint_secret
        )
    except ValueError as e:
        print("⚠️ Invalid payload:", e)
        return {"status": "invalid payload"}
    except stripe.error.SignatureVerificationError as e:
        print("❌ Invalid signature:", e)
        return {"status": "invalid signature"}

    # ✅ チェックアウト完了イベントを処理
    if event["type"] == "checkout.session.completed":
        session = event["data"]["object"]

        # クエリパラメータのdiscord_idを取得
        discord_id = session.get("client_reference_id") or session.get("metadata", {}).get("discord_id")

        if not discord_id:
            print("⚠️ discord_id not found in session")
            return {"status": "missing discord_id"}

        print(f"✅ Payment completed for Discord ID: {discord_id}")

        # Notionに決済状況を更新
        success = update_payment_status(discord_id)

        if success:
            return {"status": "success"}
        else:
            return {"status": "notion update failed"}

    return {"status": "ignored"}

File: stripeWebhook.js

import express from 'express';
import stripe from 'stripe';
import bodyParser from 'body-parser';
import dotenv from 'dotenv';
import { updatePaymentStatus } from './notionHandler.js';

dotenv.config();

const app = express();
const stripeClient = new stripe(process.env.STRIPE_SECRET_KEY);
const endpointSecret = process.env.STRIPE_WEBHOOK_SECRET;

app.post('/webhook', bodyParser.raw({ type: 'application/json' }), async (req, res) => {
  const sig = req.headers['stripe-signature'];

  let event;

  try {
    event = stripeClient.webhooks.constructEvent(req.body, sig, endpointSecret);
  } catch (err) {
    console.log(`⚠️  Webhook signature verification failed.`, err.message);
    return res.sendStatus(400);
  }

  if (event.type === 'checkout.session.completed') {
    const session = event.data.object;
    const discordId = session.client_reference_id || session.metadata?.discord_id;

    if (!discordId) {
      console.log('⚠️ discord_id not found in session');
      return res.json({ status: 'missing discord_id' });
    }

    console.log(`✅ Payment completed for Discord ID: ${discordId}`);

    const eventId = session.metadata?.event_id;
    if (!eventId) {
      console.log('⚠️ event_id not found in session metadata');
      return res.json({ status: 'missing event_id' });
    }

    const success = await updatePaymentStatus(discordId, eventId, '支払い済み');

    if (success) {
      res.json({ status: 'success' });
    } else {
      res.json({ status: 'notion update failed' });
    }
  } else {
    res.json({ status: 'ignored' });
  }
});

const PORT = process.env.PORT || 8000;
app.listen(PORT, () => console.log(`Server running on port ${PORT}`));
